clippy_lints/src/unsafe_sizeof_count_copies.rs

```diff
@@ -0,0 +1,146 @@
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, Ty, TyS, TypeAndMut};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Detects expressions where
+    /// size_of::<T> is used as the count argument to unsafe
+    /// memory copying functions like ptr::copy and
+    /// ptr::copy_nonoverlapping where T is the pointee type
+    /// of the pointers used
+    ///
+    /// **Why is this bad?** These functions expect a count
+    /// of T and not a number of bytes, which can lead to
+    /// copying the incorrect amount of bytes, which can
+    /// result in Undefined Behaviour
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust,no_run
+    /// # use std::ptr::copy_nonoverlapping;
+    /// # use std::mem::size_of;
+    ///
+    /// const SIZE: usize = 128;
+    /// let x = [2u8; SIZE];
+    /// let mut y = [2u8; SIZE];
+    /// unsafe { copy_nonoverlapping(x.as_ptr(), y.as_mut_ptr(), size_of::<u8>() * SIZE) };
+    /// ```
+    pub UNSAFE_SIZEOF_COUNT_COPIES,
+    correctness,
+    "unsafe memory copying using a byte count instead of a count of T"
+}
+
+declare_lint_pass!(UnsafeSizeofCountCopies => [UNSAFE_SIZEOF_COUNT_COPIES]);
+
+fn get_size_of_ty(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<Ty<'tcx>> {
+    match expr.kind {
+        ExprKind::Call(count_func, _func_args) => {
+            if_chain! {
+                if let ExprKind::Path(ref count_func_qpath) = count_func.kind;
+                if let Some(def_id) = cx.qpath_res(count_func_qpath, count_func.hir_id).opt_def_id();
+                if match_def_path(cx, def_id, &paths::MEM_SIZE_OF)
+                    || match_def_path(cx, def_id, &paths::MEM_SIZE_OF_VAL);
+                then {
+                    cx.typeck_results().node_substs(count_func.hir_id).types().next()
+                } else {
+                    None
+                }
+            }
+        },
+        ExprKind::Binary(op, left, right) if BinOpKind::Mul == op.node || BinOpKind::Div == op.node => {
+            get_size_of_ty(cx, left).or_else(|| get_size_of_ty(cx, right))
+        },
+        _ => None,
+    }
+}
+
+fn get_pointee_ty_and_count_expr(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<(Ty<'tcx>, &'tcx Expr<'tcx>)> {
+    if_chain! {
+        // Find calls to ptr::{copy, copy_nonoverlapping}
+        // and ptr::{swap_nonoverlapping, write_bytes},
+        if let ExprKind::Call(func, args) = expr.kind;
+        if let [_, _, count] = args;
+        if let ExprKind::Path(ref func_qpath) = func.kind;
+        if let Some(def_id) = cx.qpath_res(func_qpath, func.hir_id).opt_def_id();
+        if match_def_path(cx, def_id, &paths::COPY_NONOVERLAPPING)
+            || match_def_path(cx, def_id, &paths::COPY)
+            || match_def_path(cx, def_id, &paths::WRITE_BYTES)
+            || match_def_path(cx, def_id, &paths::PTR_SWAP_NONOVERLAPPING);
```

Comment:
```suggestion
        if FUNCTIONS.iter().any(|func| match_def_path(cx, def_id, func));
```
Given that we have already a good amount of functions to check, I think that pulling the names into a `const` array may make it easier to read.

We could do the same for the list of methods.

Reply:
I tried doing this with `const FUNCTIONS: [[&str; 3]; 8]` but it seems I need [&str; 4]'s for the slice paths:
```rs
pub const SLICE_FROM_RAW_PARTS: [&str; 4] = ["core", "slice", "raw", "from_raw_parts"];
pub const SLICE_FROM_RAW_PARTS_MUT: [&str; 4] = ["core", "slice", "raw", "from_raw_parts_mut"];
```

I tried slicing in the const context like this 
```rs
const FUNCTIONS: [&[&str]; 8] = [
    &paths::COPY_NONOVERLAPPING[..],
    &paths::COPY[..],
    &paths::WRITE_BYTES[..],
    &paths::PTR_SWAP_NONOVERLAPPING[..],
    &paths::PTR_SLICE_FROM_RAW_PARTS[..],
    &paths::PTR_SLICE_FROM_RAW_PARTS_MUT[..],
    &paths::SLICE_FROM_RAW_PARTS[..],
    &paths::SLICE_FROM_RAW_PARTS_MUT[..],
];
```
But ran into `error[E0015]: calls in constants are limited to constant functions, tuple structs and tuple variants` (I think slicing might not yet be const fn?)

Is it acceptable to just make the array not const?
```rs
    let functions: [&[&str]; 8] = [
        &paths::COPY_NONOVERLAPPING[..],
        &paths::COPY[..],
        &paths::WRITE_BYTES[..],
        &paths::PTR_SWAP_NONOVERLAPPING[..],
        &paths::PTR_SLICE_FROM_RAW_PARTS[..],
        &paths::PTR_SLICE_FROM_RAW_PARTS_MUT[..],
        &paths::SLICE_FROM_RAW_PARTS[..],
        &paths::SLICE_FROM_RAW_PARTS_MUT[..],
    ];
```

Reply:
Someone in the rust community discord pointed out relying on the array-slice coercion works in const contexts so I ended up just doing that